src/libraries/System.Linq/tests/ElementAtTests.cs

```diff
@@ -72,12 +104,426 @@ public void NullableArray_ValidIndex_ReturnsCorrectObject()
+            Assert.Equal(-1, query2[1].ElementAt(new Index(2)));
+            Assert.Equal(-1, query2[2].ElementAt(^4));
+        }
+
+        private class ArrayWrapper<T> : IList<T>
+        {
+            private readonly T[] _list;
+
+            internal ArrayWrapper(T[] list) => _list = list;
+
+            public int Count => _list.Length;
+
+            public T this[int index]
+            {
+                get => _list[index];
+                set => _list[index] = value;
+            }
+
+            public IEnumerator<T> GetEnumerator() => ((IList<T>)_list).GetEnumerator();
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+            public void CopyTo(T[] array, int arrayIndex) => _list.CopyTo(array, arrayIndex);
+
+            public bool Contains(T item) => _list.Contains(item);
+
+            public void Add(T item) => throw new NotImplementedException();
+
+            public void Clear() => throw new NotImplementedException();
+
+            public bool Remove(T item) => throw new NotImplementedException();
+
+            public bool IsReadOnly => throw new NotImplementedException();
+
+            public int IndexOf(T item) => throw new NotImplementedException();
+
+            public void Insert(int index, T item) => throw new NotImplementedException();
+
+            public void RemoveAt(int index) => throw new NotImplementedException();
+        }
+
+        [Fact]
+        public void NonEmptySource_Consistency()
+        {
+            int[] source = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+
+            Assert.Equal(5, source.ElementAt(5));
+            Assert.Equal(5, source.ElementAt(new Index(5)));
+            Assert.Equal(5, source.ElementAt(^5));
+
+            Assert.Equal(0, source.ElementAt(0));
+            Assert.Equal(0, source.ElementAt(new Index(0)));
+            Assert.Equal(0, source.ElementAt(^10));
+
+            Assert.Equal(9, source.ElementAt(9));
+            Assert.Equal(9, source.ElementAt(new Index(9)));
+            Assert.Equal(9, source.ElementAt(^1));
+
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("index", () => source.ElementAt(-1));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("index", () => source.ElementAt(^11));
+            AssertExtensions.Throws<IndexOutOfRangeException>(() => new ArrayWrapper<int>(source).ElementAt(-1));
+            AssertExtensions.Throws<IndexOutOfRangeException>(() => new ArrayWrapper<int>(source).ElementAt(^11));
+
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("index", () => source.ElementAt(10));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("index", () => source.ElementAt(new Index(10)));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("index", () => source.ElementAt(^0));
+            AssertExtensions.Throws<IndexOutOfRangeException>(() => new ArrayWrapper<int>(source).ElementAt(10));
```

Comment:
I would factor this into a separate test called something like "ElementAt surfaces IList<T> index exceptions".